src/MetadataManager/PathToHash.ts

```diff
@@ -0,0 +1,305 @@
+      const hashList = await vscode.workspace.fs.readDirectory(hashFolderUri);
+      for (const hashFile of hashList) {
+        const hashUri = vscode.Uri.joinPath(hashFolderUri, hashFile[0]);
+        let metaObj =
+            JSON.parse(Buffer.from(await vscode.workspace.fs.readFile(hashUri)).toString());
+        const hash = file[0] + hashFile[0].split('.')[0];
+        for (const path in metaObj) {
+          if (flattenMap[path] === undefined) {
+            const uri = vscode.Uri.joinPath(vscode.workspace.workspaceFolders[0].uri, path);
+            await Metadata.disable(uri, hash);
+          }
+        }
+      }
+    }
+  }
+}
+
+
+
+export class PathToHash {
+  private static _instance: PathToHash;
+  private _map: any;
+  public static startFlag: boolean = false;
+
+  private constructor() {}
+
+  /**
+   * A Singleton Object
+   *
+   * PathToHash is a data structure that stores the path of the actual data
+   * and the hash value generated by the contents of the actual data
+   * to synchronize the actual data and metadata
+   */
+  public static async getInstance() {
+    if (!this._instance) {
+      this._instance = new PathToHash();
+      this._instance._map = await this._instance.init();
+      PathToHash.startFlag = true;
+    }
+    return this._instance;
+  }
+
+  private async init() {
+    if (vscode.workspace.workspaceFolders === undefined) {
+      return;
+    }
+    const uri = vscode.workspace.workspaceFolders[0].uri;
+    const map = await this.scanRecursively(uri);
+    await MetadataSynchronizer.run(await this.getFlatMap(map));
+    return map;
+  }
+
+  /**
+   * Scan the file system and generates hash of all the files.
+   * The map is structured as a tree of directories and files.
+   *
+   * @example
+   *  uri: 'A/B/C.txt'
+   *  return: {'A': {'B': {'C.txt': <C.txt's hash value> } } } }
+   */
+  private async scanRecursively(uri: vscode.Uri) {
+    let subMap: {[key: string]: any} = {};
+    const files = await vscode.workspace.fs.readDirectory(uri);
+
+    for (const file of files) {
+      const name: string = file[0];
+      const type: vscode.FileType = file[1];
+
+      if (type === vscode.FileType.File) {
+        if (vscode.workspace.workspaceFolders !== undefined) {
+          if (isOneExplorerTargetFile(
+                  vscode.Uri.joinPath(vscode.workspace.workspaceFolders[0].uri, name))) {
+            subMap[name] = await generateHash(vscode.Uri.joinPath(uri, '/' + name));
+          }
+        }
+      } else if (type === vscode.FileType.Directory && name !== METADATA_SAVEPATH.split('/')[0]) {
+        const temp = await this.scanRecursively(vscode.Uri.joinPath(uri, '/' + name));
+        if (temp !== undefined) {
+          subMap[name] = temp;
+        }
+      }
+    }
+    if (Object.keys(subMap).length === 0) {
+      return {};
+    }
+
+    return subMap;
+  }
+
+  private async getFlatMap(map: {[key: string]: any}) {
```

Comment:
Return type of this function is not consistent too.

Reply:
I referred to the "obtainWorkspaceRoot" function of helpers and changed it to return an error instead of just returning "undefined" when there is no workspace.